pf/internal/defaults/defaults.go

```diff
@@ -0,0 +1,307 @@
+	SchemaInfos map[string]*tfbridge.SchemaInfo
+
+	// Optional. Note that ResourceInstance need not be specified for Invoke or Configure processing.
+	ResourceInstance *tfbridge.PulumiResource
+
+	// Optional. If known, these are the provider-level configuration values, to support DefaultInfo.Config.
+	ProviderConfig resource.PropertyMap
+}
+
+// Transforms a PropertyMap to apply default values specified in DefaultInfo.
+//
+// These values are specified at the bridged provider configuration level, and are applied before any Terraform
+// processing; therefore the function works at Pulumi level (transforming a PropertyMap).
+func ApplyDefaultInfoValues(ctx context.Context, args ApplyDefaultInfoValuesArgs) resource.PropertyMap {
+
+	// Can short-circuit the entire processing if there are no matching SchemaInfo entries, and therefore no
+	// matching DefaultInfo entries at all.
+	if args.SchemaInfos == nil {
+		return args.PropertyMap
+	}
+
+	t := &defaultsTransform{
+		resourceInstance: args.ResourceInstance,
+		topSchemaMap:     args.SchemaMap,
+		topFieldInfos:    args.SchemaInfos,
+		providerConfig:   args.ProviderConfig,
+	}
+	result := t.withDefaults(ctx, make(resource.PropertyPath, 0), resource.NewObjectProperty(args.PropertyMap))
+	contract.Assertf(result.IsObject(), "defaultsTransform.withDefaults returned a non-object value")
+	return result.ObjectValue()
+}
+
+func getDefaultValue(
+	ctx context.Context,
+	property resource.PropertyKey,
+	res *tfbridge.PulumiResource,
+	fieldSchema shim.Schema,
+	defaultInfo *tfbridge.DefaultInfo,
+	providerConfig resource.PropertyMap,
+) (resource.PropertyValue, bool) {
+	na := resource.NewNullProperty()
+
+	if defaultInfo == nil {
+		return na, false
+	}
+
+	if len(defaultInfo.EnvVars) != 0 {
+		for _, n := range defaultInfo.EnvVars {
+			// Following code in v3/tfbridge, ignoring set but empty env vars.
+			if str, ok := os.LookupEnv(n); ok && str != "" {
+
+				v, err := parseValueFromEnv(fieldSchema, str)
+				if err != nil {
+					msg := fmt.Errorf("Cannot parse the value of environment variable '%s'"+
+						" to populate property '%s' with a default value: %w",
+						n, string(property), err)
+					tflog.Error(ctx, msg.Error())
+					return na, false
+				}
+
+				tflog.Trace(ctx, "DefaultInfo.EnvVars applied a default from an environment variable",
+					map[string]any{"envvar": n, "property": string(property)})
+
+				return v, true
+			}
+		}
+	}
+
+	if defaultInfo.Config != "" {
+		pk := resource.PropertyKey(defaultInfo.Config)
+		if providerConfig != nil {
+			if pv, ok := providerConfig[pk]; ok {
+				tflog.Trace(ctx, "DefaultInfo.Config inherited a value from provider config",
+					map[string]any{
+						"key":      defaultInfo.Config,
+						"property": string(property),
+					})
+				return pv, true
+			}
+		}
+	}
+
+	// Unlike v3/tfbridge the code here still falls back to Value or From specs if Config did not match, this seems
+	// resaonable to do.
```

Comment:
I agree that this is reasonable for new providers, but I'm concerned this will bite us when switching over a resource from `/pkg` to `/pf`.

If the resource was depending on the old behavior, switching over will cause a diff.

Reply:
That'd require someone specifying both .Config and .Value. I think it should be super rare. But I can match the old behavior.